aptos-move/framework/aptos-framework/sources/fungible_asset.move

```diff
@@ -107,318 +132,366 @@ module aptos_framework::fungible_asset {
+        event::emit_event(&mut events.set_ungated_transfer_events, SetUngatedTransferEvent { transfer_allowed: allow });
     }
 
-    /// Burn the `amount` of fungible metadata from `account`.
-    public fun burn(
+    /// Burn the `amount` of fungible metadata from the given wallet.
+    public fun burn<T: key>(
         ref: &BurnRef,
-        wallet: &Object<FungibleAssetWallet>,
+        wallet: Object<T>,
         amount: u64
-    ) acquires FungibleAssetWallet, FungibleAssetMetadata {
-        assert!(
-            &ref.metadata == &metadata_from_wallet(wallet),
-            error::invalid_argument(EBURN_REF_AND_WALLET_MISMATCH)
-        );
+    ) acquires FungibleAssetMetadata, FungibleAssetWallet, FungibleAssetWalletEvents {
+        let metadata = ref.metadata;
+        assert!(metadata == wallet_metadata(wallet), error::invalid_argument(EBURN_REF_AND_WALLET_MISMATCH));
+        let wallet_addr = object::object_address(&wallet);
         let FungibleAsset {
             metadata,
             amount,
-        } = extract(wallet, amount);
+        } = withdraw_internal(wallet_addr, amount);
         decrease_supply(&metadata, amount);
     }
 
-    /// Withdarw `amount` of fungible asset from `wallet` by the owner.
-    public fun withdraw(
-        account: &signer,
-        wallet: &Object<FungibleAssetWallet>,
-        amount: u64
-    ): FungibleAsset acquires FungibleAssetWallet {
-        assert_owner(account, wallet);
-        extract(wallet, amount)
-    }
-
-    /// Deposit `amount` of fungible asset to `wallet`.
-    public fun deposit(
-        wallet: &Object<FungibleAssetWallet>,
-        fa: FungibleAsset,
-    ) acquires FungibleAssetWallet {
-        let FungibleAsset { metadata, amount } = fa;
-        // ensure merging the same coin
-        let wallet = borrow_fungible_asset_mut(wallet);
-        assert!(wallet.allow_ungated_transfer, error::invalid_argument(EUNGATED_TRANSFER_IS_NOT_ALLOWED));
-        assert!(wallet.metadata == metadata, error::invalid_argument(EFUNGIBLE_ASSET_AND_WALLET_MISMATCH));
-        wallet.balance = wallet.balance + amount;
-    }
-
-    /// Transfer `amount` of fungible metadata of `metadata` to `receiver`.
-    /// Note: it does not move the underlying object.
-    public fun transfer(
-        account: &signer,
-        amount: u64,
-        from: &Object<FungibleAssetWallet>,
-        to: &Object<FungibleAssetWallet>,
-    ) acquires FungibleAssetWallet {
-        let fa = withdraw(account, from, amount);
-        deposit(to, fa);
-    }
-
-    /// Withdarw `amount` of fungible metadata from `account` ignoring `allow_ungated_transfer`.
-    public fun withdraw_with_ref(
+    /// Withdraw `amount` of fungible metadata from `wallet` ignoring `allow_ungated_transfer`.
+    public fun withdraw_with_ref<T: key>(
         ref: &TransferRef,
-        wallet: &Object<FungibleAssetWallet>,
+        wallet: Object<T>,
```

Comment:
why this is `T`?

Reply:
Same as an earlier comment - I'm just keeping all functions that take Object references consistent and make them generic